libr/anal/fcn.c

```diff
@@ -2513,23 +2509,21 @@ R_API void r_anal_fcn_check_bp_use(RAnal *anal, RAnalFunction *fcn) {
 			case R_ANAL_OP_TYPE_ROR:
 			case R_ANAL_OP_TYPE_SAL:
 			case R_ANAL_OP_TYPE_SAR:
-			case R_ANAL_OP_TYPE_SHL:
 			case R_ANAL_OP_TYPE_SHR:
 			case R_ANAL_OP_TYPE_SUB:
 			case R_ANAL_OP_TYPE_XOR:
-			// op.dst is not filled for these operations, so for now, check for bp as dst looks like this; in the future it may be just replaced with call to can_affect_bp
-				if (op.opex.ptr) {
-					pos = strstr (op.opex.ptr, str_to_find);
-				}
-				else {
-					pos = NULL;
-				}
-				if (pos && pos - op.opex.ptr < 60) {
+			case R_ANAL_OP_TYPE_SHL:
+				if (can_affect_bp (anal, &op)) {
 					fcn->bp_frame = false;
 				}
 				break;
 			case R_ANAL_OP_TYPE_XCHG:
-				if (op.opex.ptr && strstr (op.opex.ptr, str_to_find)) {
+				if (op.dst && op.dst->reg && !op.dst->memref 
+				&& !strcmp (op.dst->reg->name, anal->reg->name[R_REG_NAME_BP])) {
+					fcn->bp_frame = false;
+				}
+				if (op.src[0] && op.src[0]->reg && !op.src[0]->memref
```

Comment:
What about adding this to the `can_affect_bp` function as well?

Reply:
Thank you for the review. I'll change that.